### PR TITLE
feat: configurable email and sms rate limiting

### PR DESCRIPTION
**Pull request by cstockton**
Adds two new configuration values for rate limiting the sending of emails and sms messages:

 - GOTRUE_RATE_LIMIT_EMAIL_SENT
 - GOTRUE_RATE_LIMIT_SMS_SENT

It is implemented with a simple rate limiter that resets a counter at a regular interval. The first intervals start time is set when the counter is initialized. It will be reset when the server is restarted, but preserved when the config is reloaded.

Syntax examples:
```
1.5       # Allow 1.5 events over 1 hour (legacy format)
100       # Allow 100 events over 1 hour (1h is default)
100/1h    # Allow 100 events over 1 hour (explicit duration)
100/24h   # Allow 100 events over 24 hours
100/72h   # Allow 100 events over 72 hours (use hours for days)
10/30m    # Allow 10  events over 30 minutes
3/10s     # Allow 3   events over 10 seconds
```

Syntax in ABNF to express the format as value:
```
value = count / rate
count = 1*DIGIT ["." 1*DIGIT]
rate = 1*DIGIT "/" ival
ival = ival-sec / ival-min / ival-hr
ival-sec = 1*DIGIT "s"
ival-min = 1*DIGIT "s"
ival-hr = 1*DIGIT "h"
```

This change was a continuation of https://github.com/supabase/auth/pull/1746 adapted to support the recent preservation of rate limiters across server reloads.

### CODE REVIEW

**Review by hf — APPROVED** [review]
Just remove that function that's unused and good to go!

File: internal/api/middleware_test.go [hunk 1]
```diff
@@ -185,6 +185,8 @@ func (ts *MiddlewareTestSuite) TestVerifyCaptchaInvalid() {
 	}
 }
 
+/*
```
Comment (J0) [inline internal/api/middleware_test.go:188]: Consider removing if unused

File: internal/conf/rate.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		r.Events = f
+		r.OverTime = defaultOverTime
+		return nil
+	}
+	parts := strings.Split(value, "/")
+	if len(parts) != 2 {
+		return fmt.Errorf("rate: value does not match rate syntax %q", value)
+	}
+
+	// 52 because the uint needs to fit in a float64
```
Comment (J0) [inline internal/conf/rate.go:37]: 🙏

**Review by J0 — APPROVED** [review]

File: internal/api/middleware_test.go [hunk 3]
```diff
@@ -388,6 +392,8 @@ func (ts *MiddlewareTestSuite) TestLimitHandler() {
 	require.Equal(ts.T(), http.StatusTooManyRequests, w.Code)
 }
 
+/*
+
 func (ts *MiddlewareTestSuite) TestLimitHandlerWithSharedLimiter() {
 	// setup config for shared limiter and ip-based limiter to work
```
[the remark this answers is not in the document]
Comment (cstockton) [reply internal/api/middleware_test.go]: Good catch, thank you Joel

**Pull request merged**